README.md
p5-www-rabbitmq-cluster
=======================

API modules for the RabbitMQ cluster API


WWW-RabbitMQ-Cluster/Changes
Revision history for Perl module WWW::RabbitMQ::Cluster

0.01 Fri Jun  6 14:38:49 2014
    - first commit

0.02 Sat Jun  7 21:35:49 2014
    - change the namespace to something less broad


WWW-RabbitMQ-Cluster/lib/WWW/RabbitMQ/Cluster/Shovel.pm
package WWW::RabbitMQ::Cluster::Shovel;

use strict;
use warnings;

our $VERSION = '0.01';

sub new
{
	my $class  = shift;
	my $broker = shift;

	my $self = {
		_config => ref($_[0]) ? $_[0] : {@_},
	};

	unless ($broker && ref($broker) eq 'WWW::RabbitMQ::Cluster::Broker') {
		die "ERROR: Did not receive a proper broker object.\n";
	}

	$self->{_config}{vhost} ||= '/';
	$self->{_broker} = $broker;
	return bless($self, $class);
}

sub delete
{
	my $self = shift;
	$self->{_broker}->httpMethod('DELETE')->apiCall("parameters/shovel/$self->{_config}{vhost}/$self->{_config}{name}");
	return {deleted => 1};
}

sub get
{
	my $self = shift;
	my $res  = $self->{_broker}->httpMethod('GET')->apiCall("parameters/shovel/$self->{_config}{vhost}/$self->{_config}{name}");
	return $self->{_config} = $res;
}

sub getConfig
{
	my $self = shift;
	my $get  = shift;
	$self->get;
	if ($get eq 'all') {
		return $self->{_config};
	}
	else {
		return $self->{_config}{$get};
	}
}

sub getOrPut
{
	my $self = shift;

	my $res;
	eval {
		$res = $self->{_broker}->httpMethod('GET')->apiCall("parameters/shovel/$self->{_config}{vhost}/$self->{_config}{name}");
	};

	if ($!) {
		unless ($! =~ /Invalid/) {
			die "ERROR[404]: Invalid Method\n";
		}
		my $res = $self->{_broker}->httpMethod('PUT')->apiCall(
			"parameters/shovel/$self->{_config}{vhost}/$self->{_config}{name}",
			$self->{_config},
		);
		return $self->{_config} = $res;
	}
	else {
		return $self->{_config} = $res;
	}
}

sub put
{
	my $self = shift;
	my $res = $self->{_broker}->httpMethod('PUT')->apiCall(
		"parameters/shovel/$self->{_config}{vhost}/$self->{_config}{name}",
		$self->{_config},
	);
	return $self->{_config} = $res;
}

1;

__END__

=head1 NAME

WWW::RabbitMQ::Cluster::Shovel

=head1 SYNOPSIS

	# You need to have a broker object first
	my $broker = WWW::RabbitMQ::Cluster::Broker->new(
		username => 'guest',
		password => 'guest,
		host     => 'localhost',
	);

	# Pass the broker object to the shovel with your shovel's config
	my $shovel = WWW::RabbitMQ::Cluster::Shovel->new(
		$broker,
		config => {
			component => 'shovel',
			name      => 'example_shovel',
			vhost     => 'example_vhost',
			value     => {
			        'ack-mode'          => 'on-confirm',
			        'delete-after'      => 'never',
			        'src-uri'           => 'amqp://',
			        'src-exchange'      => 'amq.direct',
			        'src-exchange-key'  => 'example_queue',
			        'dest-uri'          => 'amqp://guest:guest@remotehost.com/example_vhost',
			        'dest-exchange'     => 'amq.direct',
			        'dest-exchange-key' => 'example_queue',
			},
		},
	);

	# Get or Create the shovel. This is probably the best method to use in general for setting shovels
	# up, unless you are specifically looking for failures.
	$shovel->getOrPut;

	# Get all the configuration details for the shovel
	my $shovel_details = $shovel->getConfig;

=head1 DESCRIPTION

A simple wrapper around the RabbitMQ Shovel plugin

=head1 METHODS

=head2 new

Create a new WWW::RabbitMQ::Cluster::Broker Object

	Takes a $broker as its first argument, needs to be a valid WWW::RabbitMQ::Cluster::Broker object.

	Additional arguments will be used as part of the shovel configuration.

=head2 delete

Removes the shovel from the given broker.

=head2 get

Retrieves the shovel based on the configuration given

=head2 getConfig

Retrieves the current configuration for the shovel object, which can be updated with ->get or getOrPut

=head2 getOrPut

Safely lets you create or retrieve a shovel to minimize failing

=head2 put

Creates the shovel based on given configuration

=cut


WWW-RabbitMQ-Cluster/lib/WWW/RabbitMQ/Cluster/Broker.pm
package WWW::RabbitMQ::Cluster::Broker;

use strict;
use warnings;

our $VERSION = '0.01';

use LWP::UserAgent;
use HTTP::Request;
use JSON;
use Want;
use URI;

use WWW::RabbitMQ::Cluster::Shovel;

sub new
{
	my $class  = shift;
	my $self   = ref($_[0]) ? $_[0] : {@_};

	unless ($self->{username} and $self->{password} and $self->{host}) {
		my @missing_arguments;
		for my $key (qw/username password host/) {
			push(@missing_arguments, $key) unless $self->{$key};
		}
		die "Missing arguments: " . join(', ', @missing_arguments) . "\n";
	}

	$self->{base}   ||= 'api';   # url will look like http://localhost:15672/api/
	$self->{mode}   ||= 'GET';
	$self->{port}   ||= '15672'; # default port for rabbitmq api
	$self->{scheme} ||= 'http';  # by default https is not enabled

	$self->{uri} = URI->new("$self->{scheme}://$self->{host}:$self->{port}/");
	return bless($self, $class);
}

sub AUTOLOAD
{
	my $self = shift;
	our $AUTOLOAD;

	my ($key) = $AUTOLOAD =~ /.*::([\w_]+)/o;
	return if ($key eq 'DESTROY');
	push @{$self->{chain}}, $key;

	if (want('OBJECT') || want('VOID')) {
		return $self;
	}

	my $args = ref($_[0]) ? $_[0] : {@_};

	unshift(@{$self->{chain}}, $self->{base});
	my $url = join('/', @{$self->{chain}});
	$self->{chain} = [];
	$self->{uri}->path($url);

	return $self->_apiCall($args);
}

sub apiCall
{
	my $self   = shift;
	my $method = "$self->{base}/" . shift;
	my $args   = shift || {};
	$self->{uri}->path($method);
	return $self->_apiCall($args);
}

sub httpMethod
{
	my ($self, $method) = @_;
	$self->{mode} = $method;
	return $self;
}

sub getShovel
{
	my $self = shift;
	my $args = shift;
	return WWW::RabbitMQ::Cluster::Shovel->new($self, $args);
}

sub _apiCall
{
	my $self = shift;
	my $args = shift;

	my $ua = LWP::UserAgent->new;
	$ua->timeout(($self->{timeout} || 30));

	my $url = $self->{uri}->as_string;

	my $req = HTTP::Request->new($self->{mode} => $url);
	$req->header('Content-Type' => 'application/json; charset=UTF-8');
	$req->authorization_basic($self->{username}, $self->{password});

	my $parser = JSON->new->utf8(1);
	my $json   = $parser->encode($args);
	$req->content($json);

	my $response = $ua->request($req);
	$self->{mode} = 'GET';

	my $code    = $response->code;
	my $content = $response->content;

	if ($code == 200) {
		my $results = $parser->decode($content);
		return $results;
	}

	if ($code == 204) {
		return {success => 1};
	}

	if ($code == 401) {
		die "ERROR[401]: [username = $self->{username}, message => $content]\n";
	}

	if ($code == 404) {
		die "ERROR[404]: Not Found\n";
	}

	if ($code == 500 && ($content =~ /timeout/)){
		die "Error[500 - Timeout]: $content\n";
	}
	else {
		die "Error[$code]: [url = $url, message = $content]\n";
	}
}

1;

__END__

=head1 NAME

WWW::RabbitMQ::Cluster::Broker

=head1 SYNOPSIS

	# Make a call to a RabbitMQ API on a broker...
	my $broker = WWW::RabbitMQ::Cluster::Broker->new(
		username => 'guest',
		password => 'guest,
		host     => 'localhost',
	);

	# get an overview of the system
	my $overview = $broker->overview;

	# get all nodes in the cluster
	my $nodes = $broker->nodes;

	# get all open connections
	my $connections = $broker->connections;

	# publish a message to an exchange
	my $res = $broker->uriRequestMethod('POST')->exchanges->$vhost->$name->publish({
		payload          => "mymessage",
		payload_encoding => "string",
		properties       => {},
		routing_key      => "mykey",
	});

	# configure a shovel
	my $res = $broker->uriRequestMethod('PUT')->parameters->shovel->$vhost->$myshovel({
		value => {
			src-uri    => "amqp://",
			src-queue  => "my-queue",
			dest-uri   => "amqp://remote-server",
			dest-queue => "another-queue",
		},
	});

=head1 DESCRIPTION

A simple module that generically interacts with the RabbitMQ API

=cut
